FOUNDATIONS.md

```diff
@@ -127,34 +124,3 @@ The "recommended" IPLD format (currently DagCBOR) needs to support *at a minimum
 * A bottom type (null).
 
 **Motivation:** Convenience, really.
-
-## Non-Cyclic, Block-Local Relative Links
-
-That is, relative links that don't traverse out the back of an object. See the
```

Comment:
I think if you're going to remove this then maybe you need to re-introduce "block-local" and "relative" into the above section about Links since it's not explicitly covered but seems to be important from the perspective of the original authors of this doc. See that it's noted again at the very bottom of the section and https://github.com/ipld/specs/issues/1 is dedicated to it.

Reply:
I disagree holistically and completely that #1 is a good idea.  This is one of the things that "seem more like wishlists or are unratified", and I propose removing for that reason.

It's certainly not something that we do now, it's not something we plan on doing soon (or ever) to my knowledge, and it got a lot of pushback in that issue itself from the very beginning.  It's not even clear that it's philosophically directly compatible with our aims: it makes things *not a DAG*, and we spend an *awful* lot of time talking about the joy of DAGs.

In short, I agree completely with Juan's comments on #1 which include "I don't like that."

The only thing I particularly *like* in that issue is this comment: https://github.com/ipld/specs/issues/1#issuecomment-247906591 -- and specifically: "There are ways to represent cycles (on a data structure level) on an application level"... which is what I'd suggest we put in some sort of FAQ.  IPLD built-in link loading does not do relative links, because that breaks the model.  Applications can certainly invent a relative link concept and we'll try to help.  Fin.

Reply:
Without this feature, one can't form a link without breaking nodes into new blocks.

> In short, I agree completely with Juan's comments on #1 which include "I don't like that."

You're taking that out of context. The _full_ comment is:

> On one hand, i think this violates the idea of the DAG and will be confusing to people. It's ultimately a dag, but one typical property of DAGs is that when you traverse down an edge, you won't find parts you've already traversed. This effectively turns the DAG into a DG with the possibility of traversal cycles. I don't like that.

All the push-back is around enabling _cycles_ or links that cross block boundaries. This section clearly forbids:

1. Cross-block relative links as they'd be impossible to validate properly.
2. Cycles within a block.

Reply:
- I don't want to implement, document, or try to explain to people that we have "links" and then we have, separately "relatively, but not-cross block" links
  - because it's complex and a mouthful
  - because I don't think there exist any applications that want this but couldn't better do with with application level semantics of their own anyway
  - and because it goes against the general grain of the desire to make blocks a transparent detail as much as possible.

- I don't think *saying* "cycles within a block are forbidden" is the end of this.
  - As I commented above: declaring that that's a requirement now means that we'd have to implement a graph exploration that validates the non-existence of cycles on any content that has relative links.  That's a nontrivial add to the demands of the system in terms of implementation work that needs doing and in terms of features that need detailed documentation.  I don't think we should do this without a serious set of user stories saying we really need to, and I don't think issue text so far, even if it has number "1", has done that to a sufficient degree for how much of an ask this is.

Reply:
This all comes down to tradeoffs and we need to discuss it in terms of the tradeoff. In this case, the tradeoff is forcing arbitrary block boundries or not.

* With this requirement (and the "arbitrary IPLD path" requirement), I can always collapse an entire IPLD graph into a single block.
* Without this requirement, I can't do that. There are cases where I _need_ multiple blocks.

On the other hand, there _is_ a wart here: forbidding non-local relative links forbids the following operation:

```
  A
 / \
v   v
B-->C
```

to 

```
A + C // block 1
\   ^
 v /
  B   // block 2
```

_Without_ that rule, block boundries would be truly arbitrary. An argument here would be that: If we can't make block boundries truly arbitrary, why bother? My counter argument would be the example I gave at the top: being able to collapse an entire subdag is still useful (directory tree to a single block).

---

WRT validation,

The only validation required would be within a single block. The validation is literally: walk the graph and see if you hit a cycle, remembering all nodes you've touched. I agree that validation sucks but we'll have to validate data when importing it _anyways_ and, unless I'm missing something, this will pretty much always involve walking the graph.

Reply:
I'm in favour of not having inner Block relative links without cycles in the foundational document. I would put the information that we have about it (this discussion and #1) nicely written up into a design history document. From there we can take it out when we see the need to have it. I currently don't see a fundamental need for it.

Reply:
^ This.  I don't want to say 'no' eternally and with finality, but I think we should carry this through a process that recruits a lot more detail and design exploration before we consider it a 'yes' with finality either.